design/dsl/request_test.go

```diff
@@ -0,0 +1,190 @@
+package dsl_test
+
+import (
+	"testing"
+
+	"fmt"
+
+	"goa.design/goa.v2/design"
+	. "goa.design/goa.v2/design/dsl"
+	"goa.design/goa.v2/eval"
+)
+
+func TestRequest(t *testing.T) {
+	var RequestType = Type("Request", func() {
+		Description("Optional description")
+		Attribute("required", design.String)
+		Attribute("name", design.String, "a name")
+		Required("required")
+	})
+	//useful for this test but not others so defined here
+	commonRequestTypeAsserts := func(t *testing.T, o design.Object, ut *design.UserTypeExpr) {
+		keys := []string{}
+		for k, _ := range o {
+			keys = append(keys, k)
+		}
+		assertHasAll(t, []string{"required"}, ut.AllRequired())
+		assertHasAll(t, []string{"required", "name"}, keys)
+		assertDescription(t, "", o["required"].Description)
+		assertDescription(t, "a name", o["name"].Description)
+		assertAttributeType(t, o["name"], design.String)
+		assertAttributeType(t, o["required"], design.String)
+	}
+	cases := map[string]struct {
+		DSL    func()
+		Assert func(t *testing.T, req design.DataType)
+	}{
+		"ArrayOfRequest": {
```

Comment:
Not a big deal but it seems the logical progression in terms of test complexity would be inline then custom type then array then map (the last two can be interchanged). Might be good to order the tests in this order so that in case of failure one can look at the first ones which should hopefully be simpler to understand.